DIFM end-to-end test and page object

tomislav-developer opened the pull request:
## Changes proposed in this Pull Request

This is an end-to-end test for the new do-it-for-me feature. It goes through the flow of ordering a new site from WordPress using this option. It is using a flow since it is comprised of multiple pages that all are part of this single DIFM process. It is using the default path with no additional options chosen.

## Steps

1. Go to the DIFM landing page
2. Chose to build a new site
3. Enter all the relevant information
4. Skip the social page
5. Let the team decide on the theme
6. Leave the pages on default and proceed to checkout
7. Verify that you arrived at the checkout

Commit:
Initial DIFM test implementation

This contains a basic Page Object as well as a new test.

tomislav-developer commented on packages/calypso-e2e/src/lib/flows/difm-flow.ts:
```diff
@@ -0,0 +1,89 @@
+	 * @param {string} name The name for the new site.
+	 * @param {string} [tagline] An optional tagline.
+	 */
+	async enterOptions( name: string, tagline?: string ): Promise< void > {
```
I am not sure about optional parameters. I have added it this way, but it depends on how you are handling such cases in other tests.

dpasque replied:
I think this use of optional parameters looks great here! 🎉 

We tend to follow the policy of...
- If it's one optional parameter, just add it at the end.
- If it's more than one, make an optional `options` object as the last parameter that has all the optional params.

tom1slav replied:
🎉

dpasque commented on packages/calypso-e2e/src/lib/flows/difm-flow.ts:
```diff
@@ -0,0 +1,89 @@
+	 *
+	 * @param {Page} page The underlying page.
+	 */
+	constructor( private page: Page ) {}
```
Huh, today I learned you could use this syntax in TypeScript -- cool! 😄

dpasque commented on packages/calypso-e2e/src/lib/flows/difm-flow.ts:
```diff
@@ -0,0 +1,89 @@
+	startContainer: '.signup.is-do-it-for-me',
+
+	// Options page
+	siteName: '#siteTitle',
+	tagline: '#tagline',
+
+	// Social page
+	facebook: 'input[name=facebookUrl]',
```
Small style nit: we often include the target element type as part of the selector names. So, for example, `siteNameInput`.

The reason for that is we found it makes reading the selector usage inline a little easier as POM classes start to grow! E.g., `siteName` just by itself could refer to text on screen that is the site name, or also an input control to set that value.

tomislav-developer replied:
Nice catch! I actually was considering it for a moment but in the end, opted for the shorter version. Agree that adding the element type is much clearer. ✅

dpasque commented on packages/calypso-e2e/src/lib/flows/difm-flow.ts:
```diff
@@ -0,0 +1,89 @@
+	async clickButton( text: string ): Promise< void > {
+		const selector = selectors.button( text );
+
+		await this.page.waitForSelector( selector );
```
As a heads up: you actually don't need this here! Playwright does [all this really slick auto-waiting](https://playwright.dev/docs/actionability) before most of their actions, which helps keep the scripts a little leaner. 👌

tomislav-developer replied:
Ah did not know that learning something new every day 😄 I actually took this method from another instance. Would be a nice follow-up to clean this up on every occurrence.

Might actually be a good candidate for another PR to add this to a helper and have the implementation in one place due to its reusability. 🤔  WDYT @dpasque

dpasque replied:
>Might actually be a good candidate for another PR to add this to a helper and have the implementation in one place due to its reusability

That's a great question, and actually, one that Edwin and I have waffled back and forth on a lot!

While the DRYness of abstracting out a shared button clicking function is appealing, we've opted for now to just let each POM class decide whether it makes sense to expose a method like this or not.

The main reason for that is there may be situations where the underlying selector for a button (by text) may actually be more complicated than it is in this case. For example, we have a few places throughout Calypso where there are some nested iFrames. These would require extra handling for that frame. There are also might be places where we might have to make the selector a little more specific with a parental selector first. Because that underlying implementation might change, we figured it might be more confusing if there were some places that used a shared library method, but then other places that used POM-specific methods. This way, while more verbose, keeps a consistent pattern of deferring button handling to each POM class. 

Let me know if that doesn't make sense! 👍

tomislav-developer replied:
Makes perfect sense. Did not know about the different implementations 👍🏻

dpasque commented on packages/calypso-e2e/src/lib/flows/difm-flow.ts:
```diff
@@ -0,0 +1,89 @@
+	 * Go to first setup page.
+	 */
+	async startSetup(): Promise< void > {
+		await this.page.goto(
+			DataHelper.getCalypsoURL(
+				'/start/do-it-for-me/new-or-existing-site?flags=signup/redesigned-difm-flow'
+			)
+		);
+		await this.validateSetupPage();
+		await this.clickButton( 'Start a new site' );
```
I think I would leave the start button click out of this navigation method.

That way, if you wanted to add a test where you go down the existing-site path, you could re-use this initial navigation method! 🎉

tomislav-developer replied:
True 🙇🏻

dpasque commented on packages/calypso-e2e/src/lib/flows/difm-flow.ts:
```diff
@@ -0,0 +1,89 @@
+	 * @param {string} [tagline] An optional tagline.
+	 */
+	async enterOptions( name: string, tagline?: string ): Promise< void > {
+		await this.page.fill( selectors.siteName, name );
+		if ( typeof tagline !== 'undefined' ) {
+			await this.page.fill( selectors.tagline, tagline );
+		}
+		await this.clickButton( 'Continue' );
```
I would also do a similar thing here where I pull the button click out of this method and into its own action in the script.

The reason for that is twofold:
1. It makes this method a little more reusable in the future. For example, maybe someone wants to enter values into these fields, but then go backward in the flow, and see if state is preserved when they return.
2. It makes the testing script read as a little more explicit. 
For example, in the script, this currently ends up looking like...
```typescript
await difmFlow.enterOptions( 'Test Site', 'Just a little test' );
await difmFlow.validateSocialPage();
```
In reading that at a surface level, it's not clear how we got from entering the site options to the new social page.

Let me know if that doesn't make sense!

tomislav-developer replied:
Yep. This would have limited us in future PRs! Edited it for a more telling call in the test case itself.

dpasque reviewed (commented):
This PR is off to a great start! 😄  @tom1slav 

You do a really nice job fitting into the style and feel of our POM classes and test scripts. 👍  Your code is also very clean and easy to read! 🙌  I also think the choice here of doing a Flow class is a great choice for project under test.

I left a few style and design suggestions that should hopefully help as you continue to build out the PR. 

Let me know if you have any questions about the comments. 🙂

Commit:
Fixing any review comment and finishing test

Implemented any changes that were mentioned on the PR.
Also added the missing few steps for a finished test.

tom1slav commented:
@dpasque This is now ready for review 🎉 I have gone through the full order cycle until I get to the Checkout page where I just verify that I am there. 

I did not continue through that stage as in that case I am going further than just the onboarding but rather continuing on to the payment process. That would not be a problem but I was unsure how that would be handled on your project since the current implementation is also running on production.

Happy reviewing 💯

dpasque commented on packages/calypso-e2e/src/lib/flows/difm-flow.ts:
```diff
@@ -0,0 +1,115 @@
+	pagePickerContainer: '.signup__step .is-difm-page-picker',
+
+	// Checkout page
+	checkoutContainer: '.checkout__content',
```
I'd be curious for your perspective on a design tradeoff here. 

I often debate the best way to validate that a given navigation has landed me on my expected page.

Recently, I've taken to validating some piece of content (often a text title) _within_ the shell container element. I see two reasons for this...
1. It keeps with [Playwright's recommendation to favor user-facing selectors](https://playwright.dev/docs/selectors#prioritize-user-facing-attributes).
2. It adds an extra layer of validation to the test. Put another way, it keeps the test more honest to the user experience. A user wouldn't be aware of a shell container element, but rather the content within that. This testing approach would be more effective at catching a bug that affected the rendering of a component nested within the shell container component. 

That said, as projects are works-in-progress, I could easily see the content potentially changing more fluidly than the mark-up of the overall wrapper container. 

Thoughts?

tomislav-developer replied:
Agree, but have to admit, I did not work in the past like this. Makes the test much more understandable!

> That said, as projects are works-in-progress, I could easily see the content potentially changing more fluidly than the mark-up of the overall wrapper container.

I believe that we need to take ownership of the work we do, so changing some bits of code is kinda part of the job. We should not make our job harder than it is though so I just combined this with Edwin's [comment](https://github.com/Automattic/wp-calypso/pull/62630#discussion_r851917590) to make it cleaner and also test the title itself on every new page.

dpasque commented on test/e2e/specs/onboarding/signup__difm.ts:
```diff
@@ -0,0 +1,61 @@
+		it( 'Proceed to Checkout', async () => {
+			await difmFlow.clickButton( 'Go to Checkout' );
+			await difmFlow.validateCheckoutPage();
+		} );
```
I think we should test _ever-so-slightly deeper_ at each of these levels as part of this flow. 

I do really appreciate the simplicity of this test! It avoids over-testing a bunch of stuff that would be much better handled lower down on the testing pyramid, which is great.

When I'm doing basic happy-path testing, I try to target the most basic case that I expect the average user to do. In this case, I think we could a very simple configuration at each stage. I think that would more accurately reflect the average case, and could be done without adding too much fragility or bloat to the test.

Some examples...
1. Maybe add a single social media link? We often just use Automattic's pages (e.g. [Twitter)](https://twitter.com/automattic)
2. Select two more pages to get the 5 total allowed.

I also think we should validate that the expected item is in the cart in checkout. We have an existing POM class for the cart that should help with this! 👍

tomislav-developer replied:
Good suggestion. I added some variations to the minimum action so that every step now has some interaction on the page just like a normal user would. 👍🏻

dpasque reviewed (commented):
@tom1slav Changes look great! Your changes continue to read really cleanly and fit well with our existing tests -- nice work! 🎉 

I added a couple more follow-up comments directly related to the code. I'll respond to the overall questions on trial scope in separate comments. 👍

dpasque commented:
>This PR contains just one test. I was not sure how many you would like me to do for this trial and how many tests you would normally add to a single PR? Feel free to comment if you would like more 😸

Great question! 😄  

For this trial, let's have you add one more sub-suite/test. 👍  That should bring the total size of the coding portion to the standard size for most trials.

Normally, this current amount of code would be a perfect place to wrap up this PR.  However, for simplicity's sake for the trial, you can just add your next sub-suite/test to this PR here! That will keep everything in one place and avoid having to deal with any code dependencies from this PR.

worldomonation commented on packages/calypso-e2e/src/lib/flows/difm-flow.ts:
```diff
@@ -0,0 +1,115 @@
+	 * Validates that we've landed on the setup page.
+	 */
+	async validateSetupPage(): Promise< void > {
+		await this.page.waitForSelector( selectors.startContainer );
+	}
+
+	/**
+	 * Validates that we've landed on the options page.
+	 */
+	async validateOptionsPage(): Promise< void > {
+		await this.page.waitForSelector( selectors.siteNameInput );
+	}
+
+	/**
+	 * Validates that we've landed on the social page.
+	 */
+	async validateSocialPage(): Promise< void > {
+		await this.page.waitForSelector( selectors.facebookInput );
+	}
+
+	/**
+	 * Validates that we've landed on the design picker page.
+	 */
+	async validateDesignPage(): Promise< void > {
+		await this.page.waitForSelector( selectors.designPickerContainer );
+	}
+
+	/**
+	 * Validates that we've landed on the page picker page.
+	 */
+	async validatePagePickerPage(): Promise< void > {
+		await this.page.waitForSelector( selectors.pagePickerContainer );
+	}
+
+	/**
+	 * Validates that we've landed on the checkout page.
+	 */
+	async validateCheckoutPage(): Promise< void > {
+		await this.page.waitForSelector( selectors.checkoutContainer, { timeout: 60000 } );
+	}
```
Given that we seem to be validating a page load over 6 different pages that each only differ by a selector that is passed in, what might be an approach to reduce the amount of repetitive code?

tomislav-developer replied:
Agree, but I do think that we should not check for header text directly inside the test. I extracted the main selector call to a reusable method and just have methods for every page containing a validation call with just the page header. This way, if we ever have to change the method call behind it, there is only one place to look at.

WDYT @worldomonation?

worldomonation replied:
> I do think that we should not check for header text directly inside the test.

I do agree - the test does not need to be the one performing the check.

My suggestion was towards a unified method, something like:

```typescript
async validatePage( name: DIFMPages ): {
    ....
}
```

Where the `DIFMPages` is a type/interface of the different DIFM pages. 

Typically, I avoid making a function per action, instead preferring to have a generic function that accepts a parameter, and switch the action based on the parameter value. Even if it is implemented as calls to specific methods in the end, _I_ tend to abstract that away into a unified interface.

An example of this is in the `sidebar-component.ts` where I opted to have a generic function `navigate` instead of `visitMedia`, `visitPages`, `visitPosts`, etc.

I recognize there is no right or wrong answer here and which choice one chooses largely comes down design philosophy and preference, but I wanted to know if you felt strongly either for or against that.

tomislav-developer replied:
To be honest, I do not have a strong opinion for or against it. I do like to have less clutter inside a class but I think that this makes it not really smaller. I am concerned that it makes the class more complicated than it needs to be in my case. It would be much easier for someone that has not yet taken a look at the automation framework to see how they can change something when looking at my current implementation. 🤔

tomislav-developer replied:
I like to obey by KISS (keep it simple stupid) in most cases, but as I said I do not have a fixed opinion on this 🤷🏻‍♂️

worldomonation commented on packages/calypso-e2e/src/lib/flows/difm-flow.ts:
```diff
@@ -0,0 +1,115 @@
+	/**
+	 * Go to first setup page.
+	 */
+	async startSetup(): Promise< void > {
```
FYI, other POM classes implement the same thing with the name `visit`.

tomislav-developer replied:
Ah thanks for pointing this out to me! 🙇🏻

Commit:
Finishing up with the second test

This contains the finished first and second test.
The first had some minor cleanups.
The second was added in this commit.

tom1slav commented:
This should now cover the entire task 2. I implemented your feedback into my code and added the second test which I mentioned in my P2 post. So it's back to you for review @dpasque 🏓 😄

dpasque commented on packages/calypso-e2e/src/lib/flows/difm-flow.ts:
```diff
@@ -0,0 +1,192 @@
+	freeDesignButton: 'button[data-e2e-button=freeOption]',
+
+	//Page picker page
+	pageOption: ( text: string ) => `div.css-1wedqbn:has-text("${ text }")`,
```
Hrm, so I think that class name `.css-1wedqbn` is dynamic and programmatically generated. That is to stay, I'm not sure how reliable and stable it will stay over updates and new builds.

I took a look at the DOM for that page, and oof, we do not have a lot of good options in the existing markup! 😱 Got any recommendations for how to proceed with any area like that?

tomislav-developer replied:
I have found a better way to do this by using `:above`. I combined it with the text search and just say to click the image just above the defined page title.

dpasque replied:
Oh, this is so cool! I've been looking for a use case for those "positional selectors", but hadn't really come across one yet. I think this works really well here! Nice! 🎉

dpasque commented on packages/calypso-e2e/src/lib/flows/difm-flow.ts:
```diff
@@ -0,0 +1,192 @@
+	 *
+	 * @param {string} twitter The link to a twitter profile.
+	 */
+	async enterSocial( twitter: string ): Promise< void > {
```
A bit of nit: but I think this current method design may grow a little clunky? That is, if we wanted to add other social inputs in the future, we'd have to redo the signature here (and update all the corresponding places that called it).

I think two approaches that might grow a bit more smoothly would be either...
1. Just have a unique method for each form of input. So `enterTwitterAccount()`, and then we could add others in the future `enterFacebookAccount()`, etc.
2. Change the parameter to be an object with optional keys for each social media account. So something like...
```typescript
{
    twitter?: string;
    facebook?: string;
}
```
The method would then add whichever social media accounts were supplied. To start, we would only need to have the `twitter` key, but this would let us add future social media support without having to change the signature in a disruptive way!

Thoughts?

tomislav-developer replied:
If we just use optional parameters, we would not solve the problem of having to modify every method call in case we add a social profile input. You would have to use undefined everywere:
```
( 'twitter.com/...', undefined, 'facebook.com/...' )
```
 Instead, I opted to use an interface. That way you just expand the interface and add a check inside the method. If you want to add a social media account other than Twitter, you can then add it like:
```
{ twitter: '...', facebook: '...' }
```

tomislav-developer replied:
I did not use the multiple methods option as it would in my opinion make the flow object much less readable.

dpasque replied:
>Instead, I opted to use an interface. That way you just expand the interface and add a check inside the method. If you want to add a social media account other than Twitter, you can then add it like:

Ah yes, this is exactly what I had in mind! My apologies if I worded my comment a bit confusingly. I was imagining a single object parameter with optional properties. 👍

dpasque commented on test/e2e/specs/onboarding/signup__difm.ts:
```diff
@@ -0,0 +1,88 @@
+		} );
+
+		it( 'Proceed to Checkout', async () => {
+			await difmFlow.selectPage( 'Blog' );
+			await difmFlow.selectPage( 'Services' );
```
I would pull these two steps out into their own case so any failures would be more quickly understandable. 👍

tomislav-developer replied:
Good idea 👍🏻

dpasque commented on packages/calypso-e2e/src/lib/flows/difm-flow.ts:
```diff
@@ -0,0 +1,192 @@
+		await this.page.click( selector );
+	}
+
+	/**
+	 * Validates that we've are seeing the correct header.
+	 *
+	 * @param {string} title The title that the page should be showing.
+	 */
+	async validatePageTitle( title: string ): Promise< void > {
+		await this.page.waitForSelector( selectors.header( title ) );
+	}
```
I like this abstraction here! 🙌

dpasque reviewed (commented):
New tests look great, and they are passing for me for both mobile and desktop viewports! 👍 

I added a few more comments related to the new changes, let me know what you think!

Commit:
Cleaning up code after review

Using an interface for social accounts.
Splitting a test case for easier bug recognition.

tom1slav commented:
@dpasque Went over your feedback and suggestions and I think I found a fitting solution for the dynamic element 🤔 
Back to you for review and a possible merge 🏓

dpasque commented:
@tom1slav Your last round of changes looks great, thank you!! 🙌  That's all the comments I have, I'll defer to @worldomonation if he has any others.

Re: merge -- we usually don't do a merge as part of the trial process, so once everything is wrapped up comment-wise, you should be good here! 😄
